Trusted types Continuation

tete17 opened the pull request:
This is a continuation of the work started in https://github.com/LadybirdBrowser/ladybird/pull/5430

This time the focus is on being able to create policies via `TrustedTypePolicyFactory::create_policy` & being able to created TrustedTypes with `TrustedTypePolicy::create_(html|script|script_url)`

This unluck a good set of wpt tests in the trusted-types section.

My plan is to focus next on integrating this with the injection sinks api.

ADKaster commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp:
```diff
@@ -35,6 +37,79 @@ void TrustedTypePolicy::initialize(JS::Realm& realm)
+    if (!function.has_value()) {
+        // 1. If throwIfMissing throw a TypeError.
+        if (throw_if_missing)
+            return vm().throw_completion<JS::TypeError>(JS::ErrorType::TrustedTypesMissingCallback, trusted_type_name);
```
if it's super easier to use a JS::Error here, then manually construct the TypeError instead of using this forwarding method.

ADKaster commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp:
```diff
@@ -35,6 +37,79 @@ void TrustedTypePolicy::initialize(JS::Realm& realm)
+    }
+
+    // 4. Let args be << value >>.
+    Vector<JS::Value> args { JS::PrimitiveString::create(vm(), value) };
```
`Vector<Value>` on the stack is gc-unsafe. use RootVector

ADKaster commented on Libraries/LibWeb/TrustedTypes/TrustedScript.cpp:
```diff
@@ -0,0 +1,46 @@
+    Base::initialize(realm);
+}
+
+String TrustedScript::to_string() const
```
spec links

ADKaster commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.h:
```diff
@@ -30,12 +34,13 @@ class TrustedTypePolicy final : public Bindings::PlatformObject {
     String name() const { return m_name; }
 
     JS::ThrowCompletionOr<GC::Ref<TrustedHTML>> create_html(String const&, Vector<JS::Value> const&);
+    JS::ThrowCompletionOr<GC::Ref<TrustedScript>> create_script(String const&, Vector<JS::Value> const&);
```
these should *really* use WebIDL::ExceptionOr

ADKaster commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicyFactory.cpp:
```diff
@@ -154,6 +155,12 @@ bool TrustedTypePolicyFactory::is_html(const JS::Value& value)
 }
 
+// https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-isscript
+bool TrustedTypePolicyFactory::is_script(const JS::Value& value)
```
wrong direction for const (east vs west). but JS::Value should be passed by value anyway. it's pointer-sized.

ADKaster commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicyFactory.cpp:
```diff
@@ -154,6 +155,12 @@ bool TrustedTypePolicyFactory::is_html(const JS::Value& value)
+// https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-isscript
+bool TrustedTypePolicyFactory::is_script(const JS::Value& value)
+{
+    return value.is_object() && is<TrustedScript>(value.as_object()) && as<TrustedScript>(value.as_object()).data_is_set();
```
```
if (value.is_object()) {
   if (auto* trusted_script = as_if<TrustedScript>(value.as_object()); trusted_script && trusted_script->data_is_set())
       return true;
}
return false;
```

Avoid multiple redundant dynamic_cast expressions

ADKaster replied:
I'm also skeptical whether this "has data been set" thing makes any sense. The spec just wants you to make sure that the object really is that native type.  as in, "does it have the [[TrustedTypeData]] slot" (made-up spec language). I think just having the data as a string as it says should be fine.

ADKaster commented on Libraries/LibWeb/TrustedTypes/TrustedScriptURL.cpp:
```diff
@@ -0,0 +1,46 @@
+    Base::initialize(realm);
+}
+
+String TrustedScriptURL::to_string() const
```
spec links

tete17 replied:
Done

tete17 commented on Libraries/LibJS/Runtime/ErrorTypes.h:
```diff
@@ -36,6 +36,7 @@
     M(ClassIsAbstract, "Abstract class {} cannot be constructed directly")                                                          \
     M(ConstructorWithoutNew, "{} constructor must be called with 'new'")                                                            \
     M(Convert, "Cannot convert {} to {}")                                                                                           \
+    M(CSPDissallowsIt, "Content Security Policy blocked the creation of the policy {}")                                             \
```
Indeed my bad

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp:
```diff
@@ -30,4 +34,40 @@ void TrustedTypePolicy::initialize(JS::Realm& realm)
+
+    // 2. If allowedByCSP is "Blocked", throw a TypeError and abort further steps.
+    if (allowed_by_csp == "Blocked") {
+        return vm.throw_completion<JS::TypeError>(JS::ErrorType::CSPDissallowsIt, policy_name);
```
Sure

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.h:
```diff
@@ -12,24 +12,29 @@
 
 namespace Web::TrustedTypes {
 
+struct TrustedTypePolicyOptions {
```
Done

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicyFactory.h:
```diff
@@ -21,16 +22,35 @@ class TrustedTypePolicyFactory final : public Bindings::PlatformObject {
+        m_default_policy = policy;
+    }
+
+    const GC::Ptr<TrustedTypePolicy>& default_policy() const
```
Sure

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp:
```diff
@@ -0,0 +1,229 @@
+// https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicy-createhtml
+JS::ThrowCompletionOr<GC::Ref<TrustedHTML>> TrustedTypePolicy::create_html(String const& input, Vector<JS::Value> const& arguments)
+{
+    auto const trusted_type = create_a_trusted_type("TrustedHTML"_string, input, arguments);
```
Makes a lot of sense

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp:
```diff
@@ -0,0 +1,229 @@
+// https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicy-createhtml
+JS::ThrowCompletionOr<GC::Ref<TrustedHTML>> TrustedTypePolicy::create_html(String const& input, Vector<JS::Value> const& arguments)
+{
+    auto const trusted_type = create_a_trusted_type("TrustedHTML"_string, input, arguments);
+    if (trusted_type.is_error())
+        return trusted_type.error();
+    return trusted_type.value().get<GC::Ref<TrustedHTML>>();
```
Done

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp:
```diff
@@ -0,0 +1,229 @@
+        // 8. Let violation be the result of executing Create a violation object for global, policy, and directive on global, policy and "trusted-types"
+        // 9. Set violation’s resource to "trusted-types-policy".
+        // 10. Set violation’s sample to the substring of policyName, containing its first 40 characters.
+        // 11. Execute Report a violation on violation.
```
Indeed it wasn't that hard thanks for the reference

Lubrsi reviewed (changes requested):
Awesome! Here are some finishing touches :^)

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.h:
```diff
@@ -0,0 +1,59 @@
+
+namespace Web::TrustedTypes {
+
+using TrustedTypesVariants = WebIDL::ExceptionOr<Variant<
```
I am changing it to `GC:Root` as these types exist inside the idl but I am not 100% sure of it.

@ADKaster what does `if the generated code sees the type` mean here?

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicyFactory.cpp:
```diff
@@ -72,6 +82,55 @@ Optional<String> TrustedTypePolicyFactory::get_attribute_type(String const& tag_
+    if (!element_ns.has_value() || element_ns.value().is_empty())
+        element_ns = String { Namespace::HTML };
+
+    // FIXME: We don't have a method in ElementFactory that can give us the interface name but these are all the cases
+    // we care about in the table in get_trusted_type_data_for_attribute function
+    // 3. Let interface be the element interface for localName and elementNs.
+    String interface;
+    if (local_name == HTML::TagNames::iframe && element_ns == Namespace::HTML) {
+        interface = "HTMLIFrameElement"_string;
+    } else if (local_name == HTML::TagNames::script && element_ns == Namespace::HTML) {
+        interface = "HTMLScriptElement"_string;
+    } else {
+        interface = "Element"_string;
+    }
+
+    // 4. Let expectedType be null.
+    Optional<String> expected_type;
+
+    static Vector<Array<String, 3>> const table {
+        { "HTMLIFrameElement"_string, "srcdoc"_string, "TrustedHTML"_string },
+        { "HTMLScriptElement"_string, "innerText"_string, "TrustedScript"_string },
+        { "HTMLScriptElement"_string, "src"_string, "TrustedScriptURL"_string },
+        { "HTMLScriptElement"_string, "text"_string, "TrustedScript"_string },
+        { "HTMLScriptElement"_string, "textContent"_string, "TrustedScript"_string },
+        { "*"_string, "innerHTML"_string, "TrustedHTML"_string },
+        { "*"_string, "outerHTML"_string, "TrustedHTML"_string },
+    };
```
I hope you can understand that I take the quick exist and take advantage of the `suggestion` keyword here.

tete17 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicyFactory.cpp:
```diff
@@ -83,6 +142,135 @@ void TrustedTypePolicyFactory::initialize(JS::Realm& realm)
+        // 3. Let directive be the policy’s directive set’s directive which name is "trusted-types"
+        auto const directive = policy->get_directive_by_name(ContentSecurityPolicy::Directives::Names::TrustedTypes);
+
+        // 4. If directive’s value only contains a tt-keyword which is a match for a value 'none', set createViolation to true.
+        if (directive->value().size() == 1 && directive->value().first() == "'none'"_string)
+            create_violation = true;
+
+        // 5. If createdPolicyNames contains policyName and directive’s value does not contain a tt-keyword which is a match for a value 'allow-duplicates', set createViolation to true.
+        if (!created_policy_names.find(policy_name).is_end() && directive->value().find("'allow-duplicates'"_string).is_end())
+            create_violation = true;
```
I think [this](https://stackoverflow.com/questions/47387591/content-security-policy-with-uppercase-names-in-chrome) combined with the fact `DirectiveOperations.cpp:753` does case insensitive search is a yes to me.

I will update it

tete17 commented:
Okay I think I got all of them this time.

Heads up @Lubrsi I kind of gave up at the end and added an extra commit to migrate everything to Utf16 since I wasn't willing to go through the conflict hell it would have been to make it appear that it was utf16 from the beginning.

I also went for this since I modified classes from the previous MR so It needed a new commit no matter what I did.

Lubrsi commented:
The `LibWeb: Implement createHTML for TrustedTypePolicy` commit isn't atomic (it doesn't build if you reset to this commit). It just needs a small amendment patch to make it build by itself:
```patch
diff --git a/Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp b/Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp
index 01d1394368..6b982c8a21 100644
--- a/Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp
+++ b/Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp
@@ -12,6 +12,7 @@
 #include <LibWeb/ContentSecurityPolicy/PolicyList.h>
 #include <LibWeb/TrustedTypes/TrustedHTML.h>
 #include <LibWeb/WebIDL/AbstractOperations.h>
+#include <LibWeb/WebIDL/CallbackType.h>
 #include <LibWeb/WebIDL/ExceptionOr.h>
 
 namespace Web::TrustedTypes {
@@ -47,7 +48,7 @@ WebIDL::ExceptionOr<GC::Root<TrustedHTML>> TrustedTypePolicy::create_html(String
 }
 
 // https://w3c.github.io/trusted-types/dist/spec/#create-a-trusted-type-algorithm
-JS::ThrowCompletionOr<GC::Root<TrustedHTML>> TrustedTypePolicy::create_a_trusted_type(TrustedTypeName trusted_type_name, String const& value, GC::RootVector<JS::Value> const& arguments)
+WebIDL::ExceptionOr<GC::Root<TrustedHTML>> TrustedTypePolicy::create_a_trusted_type(TrustedTypeName trusted_type_name, String const& value, GC::RootVector<JS::Value> const& arguments)
 {
     auto& vm = this->vm();
     auto& realm = this->realm();
diff --git a/Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.h b/Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.h
index 9a7e7ebde3..95aa5bc317 100644
--- a/Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.h
+++ b/Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.h
@@ -44,7 +44,7 @@ private:
     explicit TrustedTypePolicy(JS::Realm&, String const&, TrustedTypePolicyOptions const&);
     virtual void initialize(JS::Realm&) override;
 
-    JS::ThrowCompletionOr<GC::Ref<TrustedHTML>> create_a_trusted_type(TrustedTypeName, String const&, GC::RootVector<JS::Value> const& values);
+    WebIDL::ExceptionOr<GC::Root<TrustedHTML>> create_a_trusted_type(TrustedTypeName, String const&, GC::RootVector<JS::Value> const& values);
 
     WebIDL::ExceptionOr<JS::Value> get_trusted_type_policy_value(TrustedTypeName, String const& value, GC::RootVector<JS::Value> const& values, ThrowIfCallbackMissing throw_if_missing);
 ```
This will cause a merge conflict with the `LibWeb: Implement all IDL methods that require TrustedScript`. Take the changes from that commit, the patch above is just to make that commit build by itself.

tete17 commented:
I am impressed @Lubrsi 

Did you use a tool to check the commits or was it just a visual review thing?

Anyway it is done :)

tete17 commented:
> > I am impressed @Lubrsi
> > Did you use a tool to check the commits or was it just a visual review thing?
> > Anyway it is done :)
> 
> I cherry picked each commit one by one and built every time after each cherry pick :^)

You are patient :)

Lubrsi commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicyFactory.cpp:
```diff
@@ -83,30 +143,160 @@ void TrustedTypePolicyFactory::initialize(JS::Realm& realm)
+        if (directive->value().size() == 1 && directive->value().first().equals_ignoring_ascii_case(ContentSecurityPolicy::Directives::KeywordTrustedTypes::None))
+            create_violation = true;
+
+        // 5. If createdPolicyNames contains policyName and directive’s value does not contain a tt-keyword which is a match for a value 'allow-duplicates', set createViolation to true.
+        if (!created_policy_names.find(policy_name).is_end()
+            && directive->value().find_if([](auto const& directive_value) { return directive_value.equals_ignoring_ascii_case(ContentSecurityPolicy::Directives::KeywordTrustedTypes::AllowDuplicates); }).is_end())
+            create_violation = true;
+
+        // 6. If directive’s value does not contain a tt-policy-name, which value is policyName, and directive’s value does not contain a tt-wildcard, set createViolation to true.
+        if (directive->value().find(policy_name.to_utf8()).is_end()
+            && directive->value().find_if([](auto const& directive_value) { return directive_value.equals_ignoring_ascii_case(ContentSecurityPolicy::Directives::KeywordTrustedTypes::WildCard); }).is_end())
+            create_violation = true;
```
suggestion: There's a bit too much going on in one line in both of these, breaking them down into multiple lines will make this easier to follow:
```suggestion
        // 5. If createdPolicyNames contains policyName and directive’s value does not contain a tt-keyword which is a match for a value 'allow-duplicates', set createViolation to true.
        auto created_policy_names_iterator = created_policy_names.find(policy_name);
        if (!created_policy_names_iterator.is_end()) {
            auto maybe_allow_duplicates = directive->value().find_if([](auto const& directive_value) {
                return directive_value.equals_ignoring_ascii_case(ContentSecurityPolicy::Directives::KeywordTrustedTypes::AllowDuplicates);
            });

            if (maybe_allow_duplicates.is_end())
                create_violation = true;
        }

        // 6. If directive’s value does not contain a tt-policy-name, which value is policyName, and directive’s value does not contain a tt-wildcard, set createViolation to true.
        auto directive_value_iterator = directive->value().find(policy_name.to_utf8());
        if (directive_value_iterator.is_end()) {
            auto maybe_wild_card = directive->value().find_if([](auto const& directive_value) {
                return directive_value.equals_ignoring_ascii_case(ContentSecurityPolicy::Directives::KeywordTrustedTypes::WildCard);
            });
            
            if (maybe_wild_card.is_end())
                create_violation = true;
        }
```

Lubrsi commented:
LGTM, just one more comment :^)

tete17 commented:
Last changes applied :)

Should be good to merge @Lubrsi :)

And thanks again for the patience

tcl3 commented on Libraries/LibWeb/TrustedTypes/TrustedTypePolicy.cpp:
```diff
@@ -10,6 +10,7 @@
 #include <LibJS/Runtime/Realm.h>
 #include <LibWeb/Bindings/Intrinsics.h>
 #include <LibWeb/TrustedTypes/TrustedHTML.h>
+#include <LibWeb/TrustedTypes/TrustedScript.h>
```
There are currently 2 "LibWeb: Implement TrustedScript type"  commits. It looks like these should be merged into a single commit?

tete17 replied:
With all the rebasing and edited I made I mistake

I managed to fix it let me know if it is okay now :)

tcl3 commented on Libraries/LibWeb/TrustedTypes/TrustedHTML.cpp:
```diff
@@ -0,0 +1,43 @@
+// https://w3c.github.io/trusted-types/dist/spec/#trustedhtml-stringification-behavior
+Utf16String const& TrustedHTML::to_string() const
+{
+    // 1. return the associated data value.
```
nit: We don't typically add a number to spec text that isn't numbered like this. I think most of the time for simple getters like this we just omit the text entirely.

I think the same comment applies to all `to_string()` and `to_json()` methods that currently follow this pattern.

tete17 replied:
It was so initially but I was asked by @Lubrsi to change it.

I am cool with both options :)

tcl3 replied:
Fair enough, I see the comment now. I don't feel too strongly about it.

tcl3 reviewed (commented):
Looks good! Just a couple of very minor things.

Lubrsi reviewed (approved):
LGTM, thanks for your patience and sticking through all the changes! :^)

Lubrsi commented:
Smaller MRs are typically easier to review because there's simply less to review :^)
